### ISSUE TITLE
recovery hangs forever if no write has happened

### ISSUE THREAD

**Issue by ongardie-sfdc** [salesforce/runway-model-bookkeeper#1]
Steps to repro in a simulation:
1. sendCreateLedger(1) // create ledger 1 using client 1
2. sendRecover(2, 1) // start recovery on ledger 1 using client 2

The problem is that the bookies don't know about the ledger yet. They keep no state about the ledger. But they need to prevent any future writes to the ledger before they can accept the FenceLedgerRequest message from the client.

What does BookKeeper do in this case?

**Comment by ongardie-sfdc** [comment]
@sijie @jvrao, thoughts?

**Comment by sijie** [comment]
Ah, sendRecover(2, 1) will send read request with a fence flag to read last add confirmed. On bookie side, it will first fence the ledger and read the last add confirmed. The fence request will write a fence request to journal and force the existence of a ledger. After fence succeed, it will serve the read last add confirmed. Because there is no entries, it will respond -1 as the last add confirmed.

**Comment by sijie** [comment]
@ongardie-sfdc are you running the model constantly somewhere? How do you capture this?

**Comment by ongardie-sfdc** [comment]
@sijie, sorry forgot to reply. Thanks for clarifying. I guess garbage collection is a separate issue.

> are you running the model constantly somewhere? How do you capture this?

I'm not running it actively, but you can get to it at https://runway.systems/?model=github.com/salesforce/runway-model-bookkeeper or by clicking the screenshot on the README.

**Comment by sijie** [comment]
@ongardie can you elaborate on "garbage collection is a separate issue"?

**Comment by ongardie-sfdc** [comment]
> @ongardie can you elaborate on "garbage collection is a separate issue"?

Going back to the scenario you outlined, the bookie

> ...will write a fence request to journal and force the existence of a ledger.

Will that bookie need to store that empty fenced ledger until the metadata for the ledger is deleted in ZooKeeper?

**Comment by sijie** [comment]
@ongardie-sfdc correct. the ledger' content (index file, entries) will be garbage collected after the ledger is deleted from zookeeper.
